### PR TITLE
Provide more flexibility in the SafeHandle constructor analyzer's visibility detection

### LINKED ISSUE #5231: CA1419 is too restrictive on the ctor required
### Analyzer

**Diagnostic ID**: [CA1419](https://github.com/dotnet/roslyn-analyzers/pull/5043): `Provide a public parameterless constructor for concrete types derived from 'System.Runtime.InteropServices.SafeHandle`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 6 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

### Describe the bug

CA1419 is requiring that SafeHandles have a `public` parameterless ctor.  This is too restrictive.  Most custom SafeHandles are themselves `internal` to a project, so requiring it have a `public` ctor is prohibitive; to satisfy the purpose of the rule (to be usable with a source generator that needs to be able to construct the SafeHandle), the ctor need only be as visible as the type itself, e.g. internal if internal, etc.

### Steps To Reproduce

Derive an internal type from SafeHandle and add an internal parameterless ctor.

### Expected behavior

No warning.

### Actual behavior

CA1419 fires.

cc: @jkoritzinsky

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Allow internal constructors to be used on internal SafeHandle-derived types, and allow private protected constructors to be used on private protected SafeHandle-derived types.

Fixes #5231

cc: @stephentoub

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/ProvidePublicParameterlessSafeHandleConstructor.cs [hunk 1]
```diff
@@ -73,6 +73,13 @@ private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbo
                         return;
                     }
 
+                    if ((constructor.DeclaredAccessibility == Accessibility.ProtectedAndInternal && type.DeclaredAccessibility == Accessibility.ProtectedAndInternal)
+                        || constructor.DeclaredAccessibility == Accessibility.Internal && type.DeclaredAccessibility == Accessibility.Internal)
+                    {
+                        // The parameterless constructor is as visible as the containing type, so there is no diagnostic to emit.
```
Comment (stephentoub) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/ProvidePublicParameterlessSafeHandleConstructor.cs:79]: @mavasani, is there an existing helper that does this, that says "is this member visible outside of the type"?

File: src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.md [hunk 2]
```diff
@@ -720,9 +720,9 @@ Platform compatibility analyzer requires a valid platform name and version.
 |CodeFix|False|
 ---
 
-## [CA1419](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1419): Provide a public parameterless constructor for concrete types derived from 'System.Runtime.InteropServices.SafeHandle'
+## [CA1419](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1419): Provide a parameterless constructor that is as visible as the containing type for concrete types derived from 'System.Runtime.InteropServices.SafeHandle'
 
-Providing a public parameterless constructor for a type derived from 'System.Runtime.InteropServices.SafeHandle' enables better performance and usage with source-generated interop solutions.
+Providing a parameterless constructor that is as visible as the containing type for a type derived from 'System.Runtime.InteropServices.SafeHandle' enables better performance and usage with source-generated interop solutions.
```
Comment (stephentoub) [inline src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.md:725]:
```suggestion
Providing a parameterless constructor that is at least as visible as the containing type for a type derived from 'System.Runtime.InteropServices.SafeHandle' enables better performance and usage with source-generated interop solutions.
```

**Review by stephentoub — APPROVED** [review]
Thanks!

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/ProvidePublicParameterlessSafeHandleConstructor.cs (hunk 1 above)
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/ProvidePublicParameterlessSafeHandleConstructor.cs:79]: I think you probably want to use either of the below extension methods: https://github.com/dotnet/roslyn-analyzers/blob/a0522afd52b25ad211298e308fd5bbc7734865e0/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L160-L166

Comment (jkoritzinsky) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/ProvidePublicParameterlessSafeHandleConstructor.cs:79]: I did not know those extension methods existed. I'll switch this code over to using them

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/ProvidePublicParameterlessSafeHandleConstructorTests.cs [hunk 3]
```diff
@@ -437,5 +436,68 @@ End Sub
 End Class";
             await VerifyVB.VerifyCodeFixAsync(source, source);
         }
+
+        [Fact]
+        public async Task SafeHandleDerived_WithInternalParameterlessConstructor_InternalType_NoDiagnostic_CS()
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/ProvidePublicParameterlessSafeHandleConstructorTests.cs:441]: nit: Add a work item for the relevant issue.

**Review by Youssef1313 — COMMENTED** [review]
I think this PR should target `main` since it's a false positive fix.

**Comment by jkoritzinsky** [comment]
@mavasani Which branch should this PR target? release/6.0.1xx or main?

I don't think I understand the process in this repo for determining which branch to target.

**Comment by Youssef1313** [comment]
> @mavasani Which branch should this PR target? release/6.0.1xx or main?
> 
> I don't think I understand the process in this repo for determining which branch to target.

@jkoritzinsky Please ignore my comment, the analyzer isn't in main.

For the process question, I think it is that new analyzers are added to the next release branch (to avoid breaking changes).
False positives are in `main` (but obviously if an analyzer not in `main`, then target the release branch it's in)
No clue about false negatives, probably it depends.

**Review by RikkiGibson — APPROVED** [review]

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-analysis for review

**Comment by jkoritzinsky** [comment]
I do not have permission to merge, so can someone merge this for me when it has been sufficiently reviewed/approved?

**Comment by buyaa-n** [comment]
> I do not have permission to merge, so can someone merge this for me when it has been sufficiently reviewed/approved?

@jkoritzinsky is this ready to merge?

**Comment by jkoritzinsky** [comment]
As far as I know this is ready for merge.

**Review by buyaa-n — APPROVED** [review]
LGTM merging

**Pull request merged**